ykrt/src/compile/jitc_yk/codegen/mod.rs

```diff
@@ -0,0 +1,218 @@
+//! The JIT's Code Generator.
+
+// FIXME: eventually delete.
+#![allow(dead_code)]
+
+use super::{jit_ir, CompilationError};
+use dynasmrt::ExecutableBuffer;
+#[cfg(any(debug_assertions, test))]
+use dynasmrt::{AssemblyOffset, DynasmApi};
+#[cfg(any(debug_assertions, test))]
+use std::{cell::Cell, collections::HashMap, slice};
+use typed_index_collections::TiVec;
+
+/// FIXME: one arch is hard-compiled-in, but we should be able to "cross-unit-test" different
+/// architectures.
+#[cfg(target_arch = "x86_64")]
+mod x86_64;
+
+/// Describes storage for a local variable.
+///
+/// FIXME: This assumes everything is a stack slot. When we come to implement a register allocator,
+/// this should be an enum of stack/reg.
+#[derive(Debug, Clone, Copy)]
+struct LocalAlloc {
+    /// The offset from the base pointer of the allocation.
+    ///
+    /// This is independent of which direction the stack grows. In other words, for architectures
+    /// where the stack grows downwards, you'd subtract this from the base pointer to find the
+    /// address of the allocation.
+    ///
+    /// FIXME: when we have register allocation, consider addressing stack slots by the stack
+    /// pointer, thus freeing up the base pointer for general purpose use.
+    frame_off: usize,
+}
+
+impl LocalAlloc {
+    fn new(frame_off: usize) -> Self {
+        Self { frame_off }
+    }
+}
+
+impl LocalAlloc {
+    fn frame_off(&self) -> usize {
+        self.frame_off
+    }
+}
+
+/// The result of running the code generator.
+pub(crate) struct CodeGenOutput {
+    /// The executable code itself.
+    buf: ExecutableBuffer,
+    /// Comments to be shown when printing the compiled trace using `AsmPrinter`.
+    ///
+    /// Used for testing and debugging.
+    #[cfg(any(debug_assertions, test))]
+    comments: HashMap<AssemblyOffset, Vec<String>>,
+}
+
+impl CodeGenOutput {
+    #[cfg(any(debug_assertions, test))]
+    fn disassemble(&self) -> String {
+        let tp = AsmPrinter::new(&self.buf);
+        #[cfg(any(debug_assertions, test))]
+        let tp = tp.set_comments(&self.comments);
+        tp.to_string()
+    }
+}
+
+/// The code generator.
+pub(crate) struct CodeGen<'a> {
+    jit_mod: &'a jit_ir::Module,
+    #[cfg(target_arch = "x86_64")]
+    asm: dynasmrt::x64::Assembler,
+    /// Abstract stack pointer, as a relative offset from `RBP`. The higher this number, the larger
+    /// the JITted code's stack. That means that even on a host where the stack grows down, this
+    /// value grows up.
+    asp: usize,
+    /// Maps each JIT IR local variable to where the JIT will store it at the machine level.
+    ///
+    /// Conceptually this is a hashmap, but since the keys are integers and there can never be
```

Comment:
I think it's conceptually a vector!

Reply:
It maps `InstrIdx`s to their allocations, so I consider this an optimised map.

Reply:
OK we're probably in the category of bikeshedding terminology here :)